src/approximate-qvm.lisp

```diff
@@ -0,0 +1,128 @@
+        :do (let ((simplified-kraus-ops  (reduce #'kraus-kron noise-operators)))
+              (when simplified-kraus-ops 
+                (apply-kraus-ops qvm instr simplified-kraus-ops)))))
+
+
+(defmethod transition ((qvm approx-qvm) (instr quil:gate-application))
+  ;; For the current instruction in the program, apply the gate
+  ;; corresponding to that instruction, and then apply all the kraus
+  ;; operators for all the sources of noise following that
+  ;; instruction. 
+  (let* ((gate   (pull-teeth-to-get-a-gate instr))
+        (params (mapcar (lambda (p) (force-parameter p qvm))
+                        (quil:application-parameters instr)))
+        (qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+        (t1-ops (loop :for q in qubits :collect (gethash q (t1-ops qvm))))
+        (t2-ops (loop :for q in qubits :collect (gethash q (t2-ops qvm))))
+        (kraus-ops (list t1-ops t2-ops)))
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple qubits) params)
+    (apply-all-kraus-operators qvm instr kraus-ops)
+    (incf (pc qvm))
+    qvm))
+
+
+(defmethod apply-classical-readout-noise ((qvm approx-qvm) (instr quil:measure))
+  ;; Apply the classical readout noise to the state of an APPROX-QVM
+  (%corrupt-readout qvm instr (readout-povms qvm)))
+
+
+(defun tphi (t1 t2)
+  "Calculate t_phi from T1 and T2. t_phi = (2*t1*t2) / (2*t1 + t2)"
+  (/ (* 2 (* t1 t2)) (+ t2 (* 2 t1))))
+
+
+(defun generate-damping-kraus-ops (t1 gate-time)
+  "Given a value for T1 and a GATE-TIME, generates the kraus operators corresponding to the t1 noise. "
+  (let* ((prob (/ gate-time t1)) 
+         (k0 (magicl:make-complex-matrix 2 2 (list 0 0 (sqrt prob) 0)))
+         (k1 (magicl:make-complex-matrix 2 2 (list 1 0 0  (sqrt (- 1 prob))))))
+    (list k0 k1)))
+
+
+(defun generate-dephasing-kraus-ops (t-phi gate-time)
+  "Given a value for T-PHI (dephasing time) and a GATE-TIME, calculates the kraus operators corresponding to the dephasing noise."
+  (let*  ((prob (/ gate-time t-phi))
+          (prob-k1 (/ prob 2))
+          (prob-k0 (- 1 prob-k1 ))
+          (kraus-ops (loop :for mat :in '("I" "Z")
+                           :for pj :in (list prob-k0 prob-k1)
+                           :collect (magicl:scale (sqrt pj)
+                                                  (quil:gate-matrix
+                                                   (quil:gate-definition-to-gate
+                                                    (quil:lookup-standard-gate mat)))))))
+    kraus-ops)) 
+
+
+;; TODO: Need to double check how damping dephasing kraus ops are
+;; calculated before just combining it into
+;; generate-dephasing-kraus-ops. For now I am pretty sure the
+;; calculations are the same.
+(defun generate-damping-dephasing-kraus-ops (t2 gate-time)  
+  "Calculates the kraus operators for dephasing and damping noise from T2, which is decoherence time. "
+  (generate-dephasing-kraus-ops t2 gate-time)) 
```

Comment:
One thing that's tickling my funny bone about this: here we are passing T2 through unmodified to `generate-dephasing-kraus-ops`, which claims it wants a `t-phi` as it's first arg. Are T2 and T-PHI the same thing? Or interchangeable in this case? The `tphi` function above suggests that T-PHI is some combination of T1 and T2. I'm sure this ties in to your comment above about double checking to make sure the calculations are equivalent. I just wanted to point out the t2 vs tphi thing.

Reply:
From what I understand, T-PHI is pure dephasing time for a qubit, and T2 is dephasing time with T1 (amplitude damping time) wrapped in. So both T2 and T-PHI are measures of dephasing time, but they are not equivalent. I think the way to generate the kraus operators for T2 dephasing vs T-PHI dephasing is the same calculation, but I am not sure, which is what my comment refers to above.

Reply:
Fair enough. To be clear I have no real understanding of these calculations and therefore no good reason to think they aren't equivalent; I was just pattern matching on the documented "types" of the arguments.